proposals/cel_finegrained_exceptions.md

```diff
@@ -0,0 +1,320 @@
+# Meta
+[meta]: #meta
+- Name: The CEL Fine-Grained Exceptions
+- Start Date: 2025-07-29
+- Author(s): [Mariam Fahmy](https://github.com/MariamFahmy98)
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Overview](#overview)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+
+# Overview
+[overview]: #overview
+
+This proposal outlines the design for a fine-grained exception for Kyverno's CEL-based ValidatingPolicy.
+
+The goals of this proposal are to:
+
+1. **Excluding Specific Images:** A straightforward method to exempt resources from policy enforcement based on their container image names.
+
+2. **Allowing Specific Values:** A powerful and generic mechanism to allow a resource to contain a specific configuration value (e.g., a hostPath volume, a privileged security context) that a policy would normally deny.
+
+# Motivation
+
+Currently, Kyverno's ValidatingPolicy relies on `matchConstraints.resourceRules` and `matchConstraints.excluderesourceRules` blocks to control policy scope. While effective for excluding entire resources based on metadata (labels, namespaces, names), this approach will not work for many real-world scenarios. For example:
+
+1. A policy might disallow a specific capability for all containers, but a single, trusted container within a Pod needs it. Excluding the entire Pod is not a secure option.
+
+2. A policy might enforce a set of approved volume types, but a specific application requires a `hostPath` volume.
+
+Writing complex CEL logic within the policy itself to handle these one-off cases is hard to manage, and violates the principle of separating policy definition from exceptions. This proposal introduces a dedicated, structured way to manage these fine-grained exceptions.
+
+# Proposal
+
+The main idea is to use the existing `PolicyException` resource to gather exception data and inject it into the CEL evaluation context. This makes the exception data available to the ValidatingPolicy expression through a new kyverno CEL variable.
+
+If one or more PolicyException resources match an incoming resource for a given policy, Kyverno will aggregate their specifications and populate two new context variables:
+
+1. `kyverno.excludedImages`: A list of image strings to be exempted.
```

Comment:
Should we call this `kyverno.allowedImages` to be consistent with `kyverno.allowedValues`?

Reply:
See below....maybe use `exceptions.*` instead of `kyverno.*`?

Reply:
Yes, I agree!